python-threatexchange/threatexchange/fetcher/apis/fb_threatexchange_api.py

```diff
@@ -0,0 +1,287 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+"""
+SignalExchangeAPI impl for Facebook/Meta's ThreatExchange Graph API platform.
+
+https://developers.facebook.com/programs/threatexchange
+https://developers.facebook.com/docs/threat-exchange/reference/apis/
+"""
+
+
+import typing as t
+import time
+from dataclasses import dataclass, field
+from threatexchange.fb_threatexchange.threat_updates import ThreatUpdateJSON
+from threatexchange.fetcher.simple.state import SimpleFetchDelta
+
+from threatexchange.fb_threatexchange.api import ThreatExchangeAPI, _CursoredResponse
+
+from threatexchange.fetcher import fetch_state as state
+from threatexchange.fetcher.fetch_api import SignalExchangeAPI
+from threatexchange.fetcher.collab_config import (
+    CollaborationConfigBase,
+    DefaultsForCollabConfigBase,
+)
+from threatexchange.signal_type.signal_base import SignalType
+
+
+@dataclass
+class FBThreatExchangeCollabConfig(
+    CollaborationConfigBase, DefaultsForCollabConfigBase
+):
+    privacy_group: int
+    app_token_override: t.Optional[str] = None
+
+
+@dataclass
+class FBThreatExchangeCheckpoint(state.FetchCheckpointBase):
+    """
+    State about the progress of a /threat_updates-backed state.
+
+    If a client does not resume tailing the threat_updates endpoint fast enough,
+    deletion records will be removed, making it impossible to determine which
+    records should be retained without refetching the entire dataset from scratch.
+    """
+
+    update_time: int = 0
+    last_fetch_time: int = field(default_factory=lambda: int(time.time()))
+
+    def is_stale(self) -> bool:
+        """
+        The API implementation will retain for 90 days
+
+        https://developers.facebook.com/docs/threat-exchange/reference/apis/threat-updates/
+        """
+        return time.time() - self.last_fetch_time > 3600 * 24 * 85  # 85 days
+
+    def get_progress_timestamp(self) -> int:
+        return self.update_time
+
+
+@dataclass
+class FBThreatExchangeOpinion(state.SignalOpinion):
+
+    REACTION_DESCRIPTOR_ID: t.ClassVar[int] = -1
+
+    descriptor_id: t.Optional[int]
+
+
+@dataclass
+class FBThreatExchangeIndicatorRecord(state.FetchedSignalMetadata):
+
+    opinions: t.List[FBThreatExchangeOpinion]
+
+    def get_as_opinions(  # type: ignore  # Why can't mypy tell this is a subclass?
```

Comment:
mypy can't work with this because list is an invariant type in python.

so, you can't do:

```python
b: t.List[SubClass] = []
a: t.List[SuperClass] = []

a = b # mypy implodes!
```


Two ways to handle this...

Either, use `Sequence` which is a covariant type.

Or, in the super class, redefine the return type as a typevar. 

```python
SignalOpinionOrSubclass = TypeVar('SignalOpinionOrSubclass', bound=SignalOpinion)

class FetchedSignalMetadata:
  def get_as_opinions(...) → t.List[SignalOpinionOrSubclass]:
    pass
```

Reply:
Yea, I eventually learned the reasons (in relation to Dict, which has the same relationship with Mapping).